cmd/refresh-uids-from-ferry/main.go

```diff
@@ -134,15 +142,31 @@ func initFlags() {
 	viper.SetDefault("notifications.admin_email", "fife-group@fnal.gov")
 
 	// Flags
-	pflag.StringP("configfile", "c", "", "Specify alternate config file")
-	pflag.BoolP("test", "t", false, "Test mode.  Query FERRY, but do not make any database changes")
-	pflag.Bool("version", false, "Version of Managed Tokens library")
 	pflag.String("admin", "", "Override the config file admin email")
 	pflag.String("authmethod", "tls", "Choose method for authentication to FERRY.  Currently-supported choices are \"tls\" and \"jwt\"")
+	pflag.StringP("configfile", "c", "", "Specify alternate config file")
+	pflag.Bool("disable-notifications", false, "Do not send admin notifications")
+	pflag.Bool("dont-notify", false, "Same as --disable-notifications")
+	pflag.BoolP("test", "t", false, "Test mode.  Query FERRY, but do not make any database changes")
 	pflag.BoolP("verbose", "v", false, "Turn on verbose mode")
+	pflag.Bool("version", false, "Version of Managed Tokens library")
 
 	pflag.Parse()
 	viper.BindPFlags(pflag.CommandLine)
+
+	// Aliases
+	// TODO There's a possible bug in viper, where pflags don't get affected by registering aliases.  The following should work, at least for one alias:
+	//  viper.RegisterAlias("dont-notify", "disableNotifications")
+	//  viper.RegisterAlias("disable-notifications", "disableNotifications")
+	// Instead, we have to work around this after we read in the config file (see setup())
+}
+
+// NOTE See initFlags().  This workaround will be removed when the possible viper bug referred to there is fixed.
+func disableNotifyFlagWorkaround() {
+	if viper.GetBool("disable-notifications") || viper.GetBool("dont-notify") {
```

Comment:
How exactly does this work? Seems a lil odd for there to be a method that returns a bool for the supplied ... instance variables? Just never saw that before.

Reply:
This is a great question.  I'll give you a short answer and a long one.

### Short answer:
`viper` here refers to the package `viper`.  When you call the function, `viper.GetBool`, that calls [this function](https://pkg.go.dev/github.com/spf13/viper@v1.12.0#GetBool).  The configuration here in a hidden global variable, and accessed by the `viper.Get*` functions.  This hiding and magic is done in the `viper` package itself, not anything within Managed Tokens.  That way, you can, out of the box, do something like this globally, without having to pass in your configuration to functions as a variable:

```go
func function1() {
   viper.Set("this_year", 2024)
}

func main() {
    function1()
    fmt.Println(viper.GetInt("this_year")) // prints 2024
}
```


### Long answer
This works for simpler applications like Managed Tokens, where configuration can be shared globally.  However, it does lead to confusion, because Go tends to not like "magic", even when it's following a very standard Singleton pattern, like it does here.  So people will naturally ask exactly what you did, wondering how "`viper`" knows where to get the information.

The way it works under the hood is that the `viper` library has a hidden (unexported) global variable, `v`, of type `*viper.Viper`. When you do something like `viper.Set("this_year", 2024)`, that function actually calls `v.Set("this_year", 2024)` ([literally](https://github.com/spf13/viper/blob/v1.12.0/viper.go#L1501) it's just that).  Similarly, `viper.GetInt("this_year")` just calls (effectively) `v.GetInt("this_year")`.

This allows the `viper` package to have a simple way for people to use the library out of the box.  Otherwise, you'd have to do something like:

```go
func function1(v *viper.Viper) {
   v.Set("this_year", 2024)
}

func main() {
   v := viper.New()  // https://pkg.go.dev/github.com/spf13/viper@v1.12.0#New
   function1(v)
   fmt.Println(v.GetInt("this_year")) // prints 2024
}
```

So you add that complexity of having to pass in the `*viper.Viper` instance to every place that needs configuration.  In this contrived example, it seems like it's not that much more code, but you can imagine that it'd get more verbose when you have to add this argument to each function that needs access to the configuration.  

However, while the second way of doing it is longer, critically, _it is more clear - there is no magic_.  That's why you're starting to see library authors moving away from this Singleton pattern of `viper.GetBool`, and encouraging the users to do it the longer but clearer way.

In the future, I'm actually in favor of refactoring this code to do it this more verbose/longer way (we'd also want to do it for the logging library), so there's nothing getting hidden.